### PR TITLE
Allowed to set more Environment Infos in tests

### LINKED ISSUE #349: [BC Idea]: Allow to override all properties of the EnvironmentInformation
### BC Idea Link

https://github.com/microsoft/BCApps/pull/342

### Description

This is already implemented and allow you to override all Property from the EnvironmentInformation in tests

### PR DESCRIPTION
**Pull request by jonaswre**
#### Summary
This PR allows the developer to set everything regarding your test.

This relates to https://github.com/microsoft/ALAppExtensions/pull/24899

Fixes #349 

Fixes [AB#495664](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/495664)

### CODE REVIEW

**Comment by JesperSchulz** [comment]
@jonaswre, could you create an issue? We'll approve it right away, but going forward we need approved issues before PRs are created.

**Comment by jonaswre** [comment]
> Issue #349 is not valid. Please make sure you link an **issue** that exists, is **open** and is **approved**.

@JesperSchulz

**Comment by jonaswre** [comment]
I thanks for the input. 

Since it's my code I wanted to defend it. But the only counter I have is the functionality exists for OnPrem already. 
And Can I deploy Test Library Apps in a SaaS Production Environment? 

I can set an OnPrem System to answer IsSaaS with true. So it might be more consistent to allow it either for both or none. 

For example If a partner allows free usage in SaaS Sandboxes is it something that could be exploited today. But then you can't secure any app onprem.

**Comment by TKapitan** [comment]
@jonaswre Ok, seems reasonable :)

**Review by TKapitan — APPROVED** [review]

**Review by JesperSchulz — DISMISSED** [review]
We had some more internal discussions and based on those I'm signing off on this PR. Given that the Test Library isn't installed in production, there is no way to access these internal events. The PR is hence safe.

**Review by IhorHandziuk — APPROVED** [review]

**Review by JesperSchulz — DISMISSED** [review]

**Review by JesperSchulz — APPROVED** [review]

**Comment by JesperSchulz** [comment]
Hi @jonaswre,

I've today tried to merge your changes, but your PR doesn't seem to be in a very good shape. I've fixed quite a few compilation errors, but now the tests are failing. The tests weren't in very good shape either (you can see all the changes I've made so far in the past 5 commits). I think you're forgetting to activate the binding of the events in your test. Please take one more look at your PR, make sure all messages etc. are fine and make sure your tests pass. When CI is green, this PR is ready to merge and release!

File: src/System Application/App/Environment Information/src/EnvironmentInformationImpl.Codeunit.al [hunk 1]
```diff
@@ -149,5 +193,54 @@ codeunit 3702 "Environment Information Impl."
         // An event which asks for the AppId to be filled in by the subscriber.
         // Do not use this event in a production environment. This should be subscribed to only in tests.
     end;
+
+    [InternalEvent(false)]
+    procedure OnBeforeIsSaaSInfrastructure(var TestMode: Boolean; var IsSaaSInfrastructure: Boolean)
+    begin
+        // An event which asks for the AppId to be filled in by the subscriber.
```
Comment (JesperSchulz) [inline src/System Application/App/Environment Information/src/EnvironmentInformationImpl.Codeunit.al]: Copy paste mistake. There is no AppId in this event.

File: src/System Application/App/Environment Information/src/EnvironmentInformationImpl.Codeunit.al [hunk 2]
```diff
@@ -149,5 +193,54 @@ codeunit 3702 "Environment Information Impl."
...
+        // Do not use this event in a production environment. This should be subscribed to only in tests.
+    end;
+
+    [InternalEvent(false)]
+    procedure OnBeforeVersionInstalled(var TestMode: Boolean; var VersionInstalled: Integer)
+    begin
+        // An event which asks for the AppId to be filled in by the subscriber.
```
Comment (JesperSchulz) [inline src/System Application/App/Environment Information/src/EnvironmentInformationImpl.Codeunit.al]: Copy paste mistake. There is no AppId in this event.

File: src/System Application/App/Environment Information/src/EnvironmentInformationImpl.Codeunit.al [hunk 3]
```diff
@@ -149,5 +193,54 @@ codeunit 3702 "Environment Information Impl."
...
+        // Do not use this event in a production environment. This should be subscribed to only in tests.
+    end;
+
+    [InternalEvent(false)]
+    procedure OnBeforeIsSaaS(var TestMode: Boolean; var IsSaaS: Boolean)
+    begin
+        // An event which asks for the AppId to be filled in by the subscriber.
```
Comment (JesperSchulz) [inline src/System Application/App/Environment Information/src/EnvironmentInformationImpl.Codeunit.al]: Copy paste mistake. There is no AppId in this event.

File: src/System Application/App/Environment Information/src/EnvironmentInformationImpl.Codeunit.al [hunk 4]
```diff
@@ -149,5 +193,54 @@ codeunit 3702 "Environment Information Impl."
...
+        // Do not use this event in a production environment. This should be subscribed to only in tests.
+    end;
+
+    [InternalEvent(false)]
+    procedure OnBeforeEnvironmentName(var TestMode: Boolean; var EnvironmentName: Text)
+    begin
+        // An event which asks for the AppId to be filled in by the subscriber.
```
Comment (JesperSchulz) [inline src/System Application/App/Environment Information/src/EnvironmentInformationImpl.Codeunit.al]: Copy paste mistake. There is no AppId in this event.

File: src/System Application/App/Environment Information/src/EnvironmentInformationImpl.Codeunit.al [hunk 5]
```diff
@@ -149,5 +193,54 @@ codeunit 3702 "Environment Information Impl."
...
+        // Do not use this event in a production environment. This should be subscribed to only in tests.
+    end;
+
+    [InternalEvent(false)]
+    procedure OnBeforeIsSandbox(var TestMode: Boolean; var IsSandbox: Boolean)
+    begin
+        // An event which asks for the AppId to be filled in by the subscriber.
```
Comment (JesperSchulz) [inline src/System Application/App/Environment Information/src/EnvironmentInformationImpl.Codeunit.al]: Copy paste mistake. There is no AppId in this event.

File: src/System Application/App/Environment Information/src/EnvironmentInformationImpl.Codeunit.al [hunk 6]
```diff
@@ -149,5 +193,54 @@ codeunit 3702 "Environment Information Impl."
...
+        // Do not use this event in a production environment. This should be subscribed to only in tests.
+    end;
+
+    [InternalEvent(false)]
+    procedure OnBeforeIsProduction(var TestMode: Boolean; var IsProduction: Boolean)
+    begin
+        // An event which asks for the AppId to be filled in by the subscriber.
```
Comment (JesperSchulz) [inline src/System Application/App/Environment Information/src/EnvironmentInformationImpl.Codeunit.al]: Copy paste mistake. There is no AppId in this event.

File: src/System Application/App/Environment Information/src/EnvironmentInformationImpl.Codeunit.al [hunk 7]
```diff
@@ -149,5 +193,54 @@ codeunit 3702 "Environment Information Impl."
...
+        // Do not use this event in a production environment. This should be subscribed to only in tests.
+    end;
+
+    [InternalEvent(false)]
+    procedure OnBeforeApplicationFamily(var TestMode: Boolean; var ApplicationFamily: Text)
+    begin
+        // An event which asks for the AppId to be filled in by the subscriber.
```
Comment (JesperSchulz) [inline src/System Application/App/Environment Information/src/EnvironmentInformationImpl.Codeunit.al]: Copy paste mistake. There is no AppId in this event.

File: src/System Application/Test Library/Environment Information/src/EnvironmentInfoTestLibrary.Codeunit.al [hunk 8]
```diff
@@ -36,23 +42,91 @@ codeunit 135094 "Environment Info Test Library"
...
+    /// Uses <see cref="OnBeforeVersionInstalled"/> event.
+    /// </summary>
+    /// <param name="NewVersionInstalled">The desired new Version Installed</param>
+    procedure SetVersionInstalled(NewVersionInstalled: Integer)
+    begin
+        TestVersionInstalled := NewVersionInstalled;
+    end;
+
+    /// <summary>
+    /// Sets the App ID that of the current application (for example, 'FIN' - Financials) when the subscription is bound.
```
Comment (JesperSchulz) [inline src/System Application/Test Library/Environment Information/src/EnvironmentInfoTestLibrary.Codeunit.al]: That's not what this test does.

File: src/System Application/Test Library/Environment Information/src/EnvironmentInfoTestLibrary.Codeunit.al [hunk 9]
```diff
@@ -36,23 +42,91 @@ codeunit 135094 "Environment Info Test Library"
...
+    /// Uses <see cref="OnBeforeIsSaaS"/> event.
+    /// </summary>
+    /// <param name="NewIsSaaS">The desired new Is SaaS</param>
+    procedure SetIsSaaS(NewIsSaaS: Boolean)
+    begin
+        TestIsSaaS := NewIsSaaS;
+    end;
+
+    /// <summary>
+    /// Sets the App ID that of the current environment when the subscription is bound.
```
Comment (JesperSchulz) [inline src/System Application/Test Library/Environment Information/src/EnvironmentInfoTestLibrary.Codeunit.al]: Again, none of the summaries seem to match the test.

**Review by JesperSchulz — CHANGES_REQUESTED** [review]
The PR seems to be in a very early stage with regards to code quality. There are many copy/paste mistakes and the tests don't pass.

**Comment by jonaswre** [comment]
So does that mean your internal discussions came to the conclusion this is the proper way to do it? 

If so I'll get the rest over the line.

**Comment by jonaswre** [comment]
Didn't see your comment that you would sign off. I'll fix the rest to get this merged. 

P.S.This is precisely the reason why I'm a fan of Uncle Bob's Clean Code principles. Comments at best outdated and a worst misunderstanding. - if you need comments you should refactor your code to be readable without them. But here is probably more of a policy question.

**Review by JesperSchulz — APPROVED** [review]

**Review by IhorHandziuk — APPROVED** [review]

**Comment by JesperSchulz** [comment]
Yeah, the opinions on code documentation are... diverse 😋 Personally I do like to document APIs / facades, as we do in modules these days. That way, the consumer can understand what a procedure does, without having to read the code. But when it comes to inline code documentation, I'm with Uncle Bob 😊
Anyway, almost there! There still seems to be a failing test, but once that is fixed, in this PR goes!

**Comment by jonaswre** [comment]
@JesperSchulz the failing tests are from main? But they seem to be fixed now.

**Comment by JesperSchulz** [comment]
I actually didn't check which tests were failing! Let me try to re-run the failed ones. If they fail again, I'll take a closer look! Let's get this in today 🥳

**Comment by JesperSchulz** [comment]
Tests are still failing and the failing tests are the ones introduced in this PR:

    Codeunit 135091 Environment Information Test Failure (0.058 seconds)
      Testfunction TestCanStartSessionWithTestIsolationEnabled Success (0.006 seconds)
      Testfunction TestIsSandboxIsTrueWhenTestabilitySandboxIsSet Success (0.003 seconds)
      Testfunction TestIsSandboxIsFalseWhenTestabilitySandboxIsNotSet Success (0.004 seconds)
      Testfunction TestIsSaaSIsTrueWhenTestabilitySaaSIsSet Success (0.003 seconds)
      Testfunction TestIsSaaSIsFalseWhenTestabilitySaaSIsNotSet Success (0.003 seconds)
      Testfunction TestIsSaaSInfrastructureIsTrueWhenTestabilitySaaSInfrastructureIsSetTrue Success (0.007 seconds)
  Error: Function TestVersionInstalledIsTrueWhenTestabilityVersionInstalledIsSet Invalid format of GUID string. The correct format of the GUID string is: CDEF7890-ABCD-0123-1234-567890ABCDEF where 0-9, A-F symbolizes hexadecimal digits.
      Testfunction TestVersionInstalledIsTrueWhenTestabilityVersionInstalledIsSet Failure (0.01 seconds)
        Error:
          Invalid format of GUID string. The correct format of the GUID string is: CDEF7890-ABCD-0123-1234-567890ABCDEF where 0-9, A-F symbolizes hexadecimal digits.
        Call Stack:
          Environment Information Test(CodeUnit 135091).TestVersionInstalledIsTrueWhenTestabilityVersionInstalledIsSet line 10 - System Application Test by Microsoft
          Test Runner - Mgt(CodeUnit 130454).RunTests line 21 - Test Runner by Microsoft
          Test Runner - Isol. Codeunit(CodeUnit 130450).OnRun(Trigger) line 4 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunTests line 2 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunSelectedTests line 35 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunNextTest line 20 - Test Runner by Microsoft
          Command Line Test Tool(Page 130455).RunNextTest - OnAction(Trigger) line 7 - Test Runner by Microsoft
  Error: Function TestIsSaaSIsTrueWhenTestabilityIsSaaSIsSet The binding of codeunit 135094 was unsuccessful. The codeunit has already been bound.
      Testfunction TestIsSaaSIsTrueWhenTestabilityIsSaaSIsSet Failure (0.007 seconds)
        Error:
          The binding of codeunit 135094 was unsuccessful. The codeunit has already been bound.
        Call Stack:
          Environment Information Test(CodeUnit 135091).TestIsSaaSIsTrueWhenTestabilityIsSaaSIsSet line 3 - System Application Test by Microsoft
          Test Runner - Mgt(CodeUnit 130454).RunTests line 21 - Test Runner by Microsoft
          Test Runner - Isol. Codeunit(CodeUnit 130450).OnRun(Trigger) line 4 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunTests line 2 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunSelectedTests line 35 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunNextTest line 20 - Test Runner by Microsoft
          Command Line Test Tool(Page 130455).RunNextTest - OnAction(Trigger) line 7 - Test Runner by Microsoft
  Error: Function TestEnvironmentNameIsTrueWhenTestabilityEnvironmentNameIsSet The binding of codeunit 135094 was unsuccessful. The codeunit has already been bound.
      Testfunction TestEnvironmentNameIsTrueWhenTestabilityEnvironmentNameIsSet Failure (0.004 seconds)
        Error:
          The binding of codeunit 135094 was unsuccessful. The codeunit has already been bound.
        Call Stack:
          Environment Information Test(CodeUnit 135091).TestEnvironmentNameIsTrueWhenTestabilityEnvironmentNameIsSet line 3 - System Application Test by Microsoft
          Test Runner - Mgt(CodeUnit 130454).RunTests line 21 - Test Runner by Microsoft
          Test Runner - Isol. Codeunit(CodeUnit 130450).OnRun(Trigger) line 4 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunTests line 2 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunSelectedTests line 35 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunNextTest line 20 - Test Runner by Microsoft
          Command Line Test Tool(Page 130455).RunNextTest - OnAction(Trigger) line 7 - Test Runner by Microsoft
  Error: Function TestIsProductionIsTrueWhenTestabilityIsProductionIsSet The binding of codeunit 135094 was unsuccessful. The codeunit has already been bound.
      Testfunction TestIsProductionIsTrueWhenTestabilityIsProductionIsSet Failure (0.004 seconds)
        Error:
          The binding of codeunit 135094 was unsuccessful. The codeunit has already been bound.
        Call Stack:
          Environment Information Test(CodeUnit 135091).TestIsProductionIsTrueWhenTestabilityIsProductionIsSet line 3 - System Application Test by Microsoft
          Test Runner - Mgt(CodeUnit 130454).RunTests line 21 - Test Runner by Microsoft
          Test Runner - Isol. Codeunit(CodeUnit 130450).OnRun(Trigger) line 4 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunTests line 2 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunSelectedTests line 35 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunNextTest line 20 - Test Runner by Microsoft
          Command Line Test Tool(Page 130455).RunNextTest - OnAction(Trigger) line 7 - Test Runner by Microsoft
  Error: Function TestApplicationFamilyIsTrueWhenTestabilityApplicationFamilyIsSet The binding of codeunit 135094 was unsuccessful. The codeunit has already been bound.
      Testfunction TestApplicationFamilyIsTrueWhenTestabilityApplicationFamilyIsSet Failure (0.007 seconds)
        Error:
          The binding of codeunit 135094 was unsuccessful. The codeunit has already been bound.
        Call Stack:
          Environment Information Test(CodeUnit 135091).TestApplicationFamilyIsTrueWhenTestabilityApplicationFamilyIsSet line 3 - System Application Test by Microsoft
          Test Runner - Mgt(CodeUnit 130454).RunTests line 21 - Test Runner by Microsoft
          Test Runner - Isol. Codeunit(CodeUnit 130450).OnRun(Trigger) line 4 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunTests line 2 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunSelectedTests line 35 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunNextTest line 20 - Test Runner by Microsoft
          Command Line Test Tool(Page 130455).RunNextTest - OnAction(Trigger) line 7 - Test Runner by Microsoft

**Comment by jonaswre** [comment]
@JesperSchulz I think i've fixed the issue. There was only one test failing but since it failed it didn't unbind the cu correctly. So I've moved them to the triggers. And the the first issue in the version override it doesn't matter what guid you provide it just need to be a correctly formated one.

https://github.com/jonaswre/BCApps/blob/b461a277bcdf9adc6460b4a33f0a301597349d5e/src/System%20Application/Test%20Library/Environment%20Information/src/EnvironmentInfoTestLibrary.Codeunit.al#L55C39-L55C39

Should that also be a guid? It might cause error down the line if you don't provide a actual guid here.

**Comment by JesperSchulz** [comment]
Let's run it once more. If it passes, in it goes 😊

**Comment by jonaswre** [comment]
@JesperSchulz 

> Let's run it once more. If it passes, in it goes 😊 About the SetAppId function: ideally it should have been a guid I guess, but we've got quite a few tests which just add some characters to that function. I don't really feel like redoing all of those tests at this point. Let's leave it as is for now.

One more time... The trigger are not allowed in test units. I've reverted to bind/unbind in every test. But I handle it gracefully. So that binding/unbinding wont be the issue if the test fail. I think there should be a better way to do this but I currently can't think of one.

**Comment by JesperSchulz** [comment]
Now I had a chance to look closer into the CI error, which is:

  Error: Function TestFinancialsSurveyDisabledInSandbox Assert.IsTrue failed. Survey is activated.
      Testfunction TestFinancialsSurveyDisabledInSandbox Failure (0.014 seconds)
        Error:
          Assert.IsTrue failed. Survey is activated.
        Call Stack:
          Library Assert(CodeUnit 130002).IsTrue line 3 - Library Assert by Microsoft
          Satisfaction Survey Tests(CodeUnit 138074).VerifySurveyDeactivated line 2 - System Application Test by Microsoft
          Satisfaction Survey Tests(CodeUnit 138074).TestFinancialsSurveyDisabledInSandbox line 15 - System Application Test by Microsoft
          Test Runner - Mgt(CodeUnit 130454).RunTests line 21 - Test Runner by Microsoft
          Test Runner - Isol. Codeunit(CodeUnit 130450).OnRun(Trigger) line 4 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunTests line 2 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunSelectedTests line 35 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunNextTest line 20 - Test Runner by Microsoft
          Command Line Test Tool(Page 130455).RunNextTest - OnAction(Trigger) line 7 - Test Runner by Microsoft
      Testfunction TestFinancialsSurveyDisabledOnMobileDevice Success (0.004 seconds)
  Error: Function TestFinancialsSurveyDisabledOnPrem Assert.IsTrue failed. Survey is activated.
      Testfunction TestFinancialsSurveyDisabledOnPrem Failure (0.007 seconds)
        Error:
          Assert.IsTrue failed. Survey is activated.
        Call Stack:
          Library Assert(CodeUnit 130002).IsTrue line 3 - Library Assert by Microsoft
          Satisfaction Survey Tests(CodeUnit 138074).VerifySurveyDeactivated line 2 - System Application Test by Microsoft
          Satisfaction Survey Tests(CodeUnit 138074).TestFinancialsSurveyDisabledOnPrem line 15 - System Application Test by Microsoft
          Test Runner - Mgt(CodeUnit 130454).RunTests line 21 - Test Runner by Microsoft
          Test Runner - Isol. Codeunit(CodeUnit 130450).OnRun(Trigger) line 4 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunTests line 2 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunSelectedTests line 35 - Test Runner by Microsoft
          Test Suite Mgt.(CodeUnit 130456).RunNextTest line 20 - Test Runner by Microsoft
          Command Line Test Tool(Page 130455).RunNextTest - OnAction(Trigger) line 7 - Test Runner by Microsoft

Turns out, this still is a side effect of the added tests. Somehow the system still thinks the environment settings are different, after your tests have run. The survey tests depend on the system being production, SaaS, etc. but after your tests have run, the system believes it's onprem. Maybe the system needs to roll back after every test, to really lose its state again? Need to further dig into that.

**Comment by JesperSchulz** [comment]
Tip: This is most likely related to the fact that the codeunit is SingleInstance...

**Comment by jonaswre** [comment]
@JesperSchulz I had a chance to look at the issue now. The Issue has nothing to do with the single instance. 

The issue was that the EnvironmentTestLibrary was is bound in the initilize function. The Binding will set TestMode to true. That will cause the old way of overriding to be ignored. I've check the rest of the test. This is the only place where the EnvironmentTestLibrary is bound. The remaining test will function like they used to.

**Comment by JesperSchulz** [comment]
Alright! Let's run this once more! Would be great if it now passes. Then we'll merge!

**Review by JesperSchulz — APPROVED** [review]

**Comment by jonaswre** [comment]
@JesperSchulz now it failed because there is a new requirement for lowercasing if...

**Review by IhorHandziuk — APPROVED** [review]

**Review by JesperSchulz — APPROVED** [review]

**Comment by JesperSchulz** [comment]
Hi @jonaswre,

unfortunately we need to revert this PR 😟 During our final test runs of the 2024 wave 1 release, we identified a few issues with these new capabilities. I am happy to explain those issues to you, if you reach out to me through a PM in Vivia Engage.

I apologize for the lost work - unfortunately we did not catch the issues during the code reviewing phase.

**Pull request merged**